### PR TITLE
Add shape specialization to d2d geometry generation.

### PR DESCRIPTION
**Pull request by smmalis37**
Working on #309. With this change my time spent in path_from_shape (now also geometry_from_shape) is cut roughly in half.

Note that this currently builds on top of #307. Only the last 2 commits are new. That PR should be merged first, and then this rebased on top of that merge (and then given a fresh CI run), before merging.

### CODE REVIEW

File: piet-direct2d/src/lib.rs [hunk 1]
```diff
@@ -109,45 +109,54 @@ impl<'b, 'a: 'b> D2DRenderContext<'a> {
...
-                if need_close {
-                    sink.end_figure(false);
+) -> Result<Geometry, Error> {
+    // TODO: Do something special for line?
+    if let Some(rect) = shape.as_rect() {
+        Ok(d2d.create_rect_geometry(rect)?.into())
+    } else if let Some(round_rect) = shape.as_rounded_rect() {
+        Ok(d2d.create_round_rect_geometry(round_rect)?.into())
+    } else if let Some(circle) = shape.as_circle() {
+        Ok(d2d.create_circle_geometry(circle)?.into())
```
Comment (smmalis37) [inline piet-direct2d/src/lib.rs:124]: I'm open to alternative ideas on how to achieve the upcast here.

**Comment by smmalis37** [comment]
This is gonna fail CI for now because #307 is failing CI.

**Comment by smmalis37** [comment]
Now ready for review.

File: piet-direct2d/src/d2d.rs [hunk 2]
```diff
@@ -146,6 +155,30 @@ impl From<Error> for piet::Error {
     }
 }
 
+impl From<PathGeometry> for Geometry {
```
Comment (smmalis37) [inline piet-direct2d/src/d2d.rs:158]: I'm open to other ideas on cleaner/shorter ways to achieve this upcast.

Comment (raphlinus) [reply piet-direct2d/src/d2d.rs:158]: I think this is ok. From a performance perspective, upcasts in COM are basically free, but downcasts require QueryInterface.

**Review by raphlinus — APPROVED** [review]
This patch looks good to me. My understanding is that it will reduce the geometry creation overhead for clips, but no expected gain on fills or strokes, as they already have specialized methods. And for clips, I'm not expecting huge gains, because I expect more overhead in the layer creation and compositing. There's a bigger win for moving clips of rects to the axis-aligned specialization, but as discussed on the other issue, that won't help if the clip itself is a rounded rect.

Regarding the TODO, clipping using Line makes just as little sense as filling, so there's no need to do anything special in that case.

File: piet-direct2d/src/d2d.rs (hunk 2 above)
Comment (smmalis37) [reply piet-direct2d/src/d2d.rs:158]: Yeah, I just meant in terms of lines of code.

**Comment by smmalis37** [comment]
Your estimation of where the gains are is accurate, but in my case its still a decently sized win.

File: piet-direct2d/src/d2d.rs (hunk 2 above)
Comment (raphlinus) [reply piet-direct2d/src/d2d.rs:158]: By COM standards involving any kind of subclassing, I've seen a lot worse :)

**Pull request merged**